src/libstd/collections/hash/map.rs

```diff
@@ -1451,6 +1518,47 @@ impl<K, V, S> HashMap<K, V, S>
     }
 }
 
+impl<K, V, S> HashMap<K, V, S>
+    where K: Eq + Hash,
+          S: BuildHasher
+{
+    /// Creates a raw entry builder for the HashMap.
+    ///
+    /// Raw entries provide the lowest level of control for searching and
+    /// manipulating a map. They must be manually initialized with hash and
+    /// then manually searched. After this, insertions into the entry also
+    /// still require an owned key to be provided.
+    ///
+    /// Raw entries are useful for such exotic situations as:
+    ///
+    /// * Hash memoization
+    /// * Deferring the creation of an owned key until it is known to be required
+    /// * Using a HashMap where the key type can't or shouldn't be hashed and/or compared
+    ///
+    /// Unless you are in such a situation, higher-level and more foolproof APIs like
+    /// `entry` should be preferred.
+    #[unstable(feature = "raw_entry", issue = "42069")]
+    pub fn raw_entry(&mut self) -> RawEntryBuilder<K, V, S> {
+        self.reserve(1);
+        RawEntryBuilder { map: self }
+    }
+
+    /// Creates a raw immutable entry builder for the HashMap.
+    ///
+    /// This is useful for
+    /// * Hash memoization
+    /// * Querying a HashMap where the key type can't or shouldn't be hashed and/or compared
+    ///
+    /// Unless you are in such a situation, higher-level and more foolproof APIs like
+    /// `entry` should be preferred.
+    ///
+    /// Immutable raw entries have very limited use; you might instead want `raw_entry`.
+    #[unstable(feature = "raw_entry", issue = "42069")]
+    pub fn raw_entry_immut(&self) -> RawImmutableEntryBuilder<K, V, S> {
```

Comment:
Naming-wise could this perhaps be `raw_entry` and the one above be `raw_entry_mut`?

Reply:
Yeah I'm split on it

`raw_entry`/`raw_entry_mut` is absolutely more idiomatic, but `entry` matches `raw_entry_mut` and `raw_entry_mut` is the really important one.

I'm definitely being "weird" here, and am willing to relent if anyone feels strong about it.